## About Dataset

The data was sourced from https://usafacts.org and focuses on firearm-related fatalities in the United States spanning from 1981 to 2021. This dataset holds significance as it allows for an examination of the necessity for gun-related laws and the common belief that having firearms at home provides a sense of security.

Exploring this dataset presents an opportunity to delve into the factors contributing to firearm incidents and the  causes behind them.This will helps us in fostering good relationship with the soceity. 

#### Data Description

Most variables are named in quite obvious ways.

Description of the variables:
- Years: The data is for which year
- Firarm Deaths: Total number of firearms deaths in that year
- Homicide: Number of deaths because of Homicide
- Suicide: Number of deaths because of Suicide
- Unintentional: Number of deaths because of Unintentional
- Legal Inetervention/Operations of War: Deaths because of Legal Internventions
- Undetermined: Deaths for which we don't know the reason for the death
- Under 19: Number of deaths where the victim was under19
- 20-34: Age is under 20-34
- 35-49: Age is under 35-49
- 50-64: Age is under 50-64
- 65+: Age is over 65+
- Male: Number of firearm death for the male gender
- Female: Number of firearm deaths for the female gender
- State Wise Count: Firearm deaths for each state across each year

### Import Libraries

In [1]:
import pandas as pd
import altair as alt

### Read Dataset

In [2]:
firearm= pd.read_csv('firearm_deaths_usafacts (1).csv')

In [3]:
# Store the year in pandas year format
firearm['Years']=pd.to_datetime(firearm['Years'],format='%Y').dt.year

# Basic Graphs

### Graph 1: Firearm Deaths Over Perios Of Time

##### Purpose of this plot

Analyzing firearm deaths over time is critical to know. Firstly, understanding how firearm-related deaths are distributed across various causes provides valuable insights into public health and safety. It allows us to identify trends, patterns, and potential areas for intervention.

In [4]:
# Code to plot the total fireaarm deaths
total_firearm_time = alt.Chart(firearm).mark_line().encode(
    alt.X('Years:O',
           axis = alt.Axis(labelAngle=0,title='Year',labelFontSize = 10,
                         titleFontSize = 12,
                         values=[1980, 1985, 1990, 1995, 2000, 2005, 2010, 2015, 2020] )),
    alt.Y('Firearmdeaths:Q',axis = alt.Axis(title='Number Of Firearm Deaths',labelFontSize = 15,
                         titleFontSize = 12)),
    tooltip =['Years:O','Firearmdeaths:Q']
).properties(
    title=('Firearm Deaths Over Period: 1981 - 2021')
)

total_firearm_time.save('chart1.html')


#### Observation:
- In 1992, there was a notable peak in firearm deaths, followed by a gradual decrease in subsequent years. However, in 2021, there is a troubling increase in the crime rate, which is a matter of significant concern. 

It becomes important to investigate the types of firearm-related fatalities and consider potential regulations or interventions aimed at reducing the current increase in crime rates. This analysis holds the promise of shedding light on effective strategies to enhance public safety and reduce firearm-related incident. 




#### Effectiveness of the Graph:

The graph above displays the total number of firearm deaths over the period from 1981 to 2021. The x-axis represents the years, ranging from 1981 to 2021, and the y-axis represents the count of firearm deaths. The line plot allows for a clear visualization of the trend in firearm deaths over these years. It is evident that there is a noticeable increase in firearm deaths over time, particularly after the mid-2000s. This graphic effectively illustrates the rise in firearm deaths, providing a straightforward and impactful visualization that directly addresses the question of how firearm deaths have evolved over teh period. 



<BR><BR>
This gives motivation to visualize the firearm deaths by different causes over time. For instance, if there's a significant rise in firearm deaths due to a specific cause, it would prompt further investigation and potentially influence policy decisions aimed at reducing fatalities

### Graph 2: Firearm Deaths Over Period Of Time With Cause

In [5]:
## Cause of Firearm Deaths Over Time
cause_firearm_time = alt.Chart(firearm).transform_fold(
    ['Homicide', 'Suicide', 'Unintentional',
       'LegalIntervention', 'Undetermined'],
    as_=['Cause','Count']

).mark_line( ).encode(
    alt.X('Years:O',
           axis = alt.Axis(labelAngle=0,title='Year',titleFontSize = 12,
                         values=[1980, 1985, 1990, 1995, 2000, 2005, 2010, 2015, 2020])),
    alt.Y('Count:Q',
           axis = alt.Axis(title='Number Of Firearm Deaths')),
    alt.Color('Cause:N'),
    text=alt.Text('Cause:N'),
    tooltip =['Years:O','Cause:N','Count:Q']
    
).properties(
    title=('Firearm Deaths By Cause'),
    height = 400
)



cause_firearm_time.save('cause.html')
cause_firearm_time

alt.Chart(...)

#### Observations:

- The widespread of firearm-related suicides is alarmingly high and appears to be on the rise. This raises a critical point for consideration; individuals often acquire firearm licenses for self-protection, yet the majority of firearm-related deaths are attributed to suicide. This contradiction calls for a thorough reevaluation of policies regarding firearm issuance.

- Homicide stands as another significant cause of firearm-related fatalities, warranting serious concern. Implementing stricter regulations or reconsidering the issuance of firearms might contribute to a reduction in homicides.

- Notably, firearm deaths due to reasons categorized as Undetermined/Unintentional and Legal Intervention/Operations of War are considerably lower compared to Homicide and Suicide-related firearm fatalities.

- A notable reduction in Homicide deaths was observed around 1998, persisting for a considerable number of years but now again it's on rise

- However, there was a subsequent decline in suicide rates, although it was relatively insignificant. Starting from 2007, suicide rates began to rise again, reaching a peak in 2021. This alarming trend necessitates a comprehensive reevaluation of strategies to address and mitigate firearm-related suicides.

#### Effectiveness of the Graph:

The graph titled "Firearm Deaths By Cause" illustrates the distribution of firearm deaths across different causes (Homicide, Suicide, Unintentional, Legal Intervention, and Undetermined) over the years from 1981 to 2021. Each line represents a specific cause, showing the count of firearm deaths assigned to that particualar cause for each year. The x-axis represents the years, ranging from 1981 to 2021, while the y-axis represents the count of firearm deaths. The different colors distinguish between the causes. This graph effectively visualizes the variations in firearm deaths across causes, allowing for a comparative analysis of the impact of each cause on firearm-related fatalities. It provides a clear and informative representation of how different causes contribute to the overall trend in firearm deaths over the specified time period.



Now, let's generate a visualization to illustrate the trend during the years with the minimum and maximum firearm death counts. During the year 2000, we observed the lowest firearm death count, while in 2021, it reached its peak.


### Graph3: Firearm deaths across multiple age group

In [6]:
age = alt.Chart(firearm).mark_bar().transform_fold(
    ['Under19', '20-34', '35-49', '50-64', '65+'],
    as_=['Age', 'Count']
).transform_aggregate(
    sum_count='sum(Count)',
    groupby=['Age']
).encode(
    alt.X('sum_count:Q',
          axis =alt.Axis(title='Number of Firearm Deaths',
                         labelFontSize = 15,
                         titleFontSize = 15)),
    alt.Y('Age:N', 
          sort= ['Under19', '20-34', '35-49', '50-64', '65+'],
          axis = alt.Axis(labelFontSize=15,
                          title='Age Group',titleFontSize=15),
          ),
    tooltip=['sum_count:Q', 'Age:N'],
    
).properties(
    width=500,
    height=300,
    title = alt.Title(text='Firearm Death By Age Group',fontSize=20)
)
age.save('age.html')

#### Observations:
- The age group ranging from 20-34 exhibits the highest count of firearm deaths in the dataset, followed by the 35-49 age group, 50-64 age group, and then 65+ age group.
- Conversely, the age group Under19 experiences the lowest count of firearm deaths.


### Effectiveness of the Graph
The graph titled "Firearm Deaths By Age Group" presents a bar chart illustrating the distribution of firearm deaths across various age groups. Each bar represents a specific age group (Under 19, 20-34, 35-49, 50-64, and 65+), showing the total count of firearm deaths for that age group. The x-axis represents the number of firearm deaths, while the y-axis represents the age groups. The bars are sorted by age group ordered as Under19 being on the top, followed by 20-34, 35-49 and so on in increasing order. This graph effectively visualizes the proportion of firearm-related fatalities in different age categories, providing a clear and informative representation of the impact of age on firearm deaths.

<BR><BR>

### Graph 4: What's the distribution of firearm deaths across different gender over period of time?     

<BR>
<B>Motivation behind this graph:</B> Understanding the distribution of firearm deaths across different genders is crucial in comprehending the gender-specific impact of firearms. By visualizing this data, we aim to shed light on how firearm-related fatalities are distributed among various gender groups. This information can highlight disparities in firearm deaths and contribute to informed discussions and potential policy interventions aimed at reducing firearm fatalities and promoting public safety. 


In [23]:
gender_firearm_time = alt.Chart(firearm).transform_fold(
    ['Male','Female'],as_=['Gender','Count']

).mark_line( ).encode(
    alt.X('Years:O',
          axis = alt.Axis(labelAngle=-45)),
    alt.Y('Count:Q'),
    alt.Color('Gender:N'),
    tooltip = ['Years:O','Count:Q','Gender:N']
    
).properties(
    title=('Firearm deaths Across Years By Gender 1981-2021')
)
gender_firearm_time.save('gender.html')
gender

NameError: name 'gender' is not defined

#### Observations:

- There's a noticeable disparity in firearm-related deaths based on gender. The minimum difference between male and female firearm deaths is approximately 20,000, raising serious concerns.
- Moreover, male firearm-related deaths have been increasing at a significantly higher rate compared to female firearm-related deaths since 2014.


#### Effectiveness of the Graph:

The graph titled "Firearm Deaths Across Years by Gender (1981-2021)" presents a line chart illustrating the trend of firearm deaths over the years for both genders, Male and Female. The x-axis represents the years from 1981 to 2021, and the y-axis represents the count of firearm deaths. Each line corresponds to a gender (Male or Female) and shows how the number of firearm deaths has changed over the years for that gender. The chart effectively allows comparison of firearm deaths by gender over the specified time period, providing insights into patterns in firearm-related fatalities.

Now, Let's visualize the fiream deaths across different states in the US from 1981-2021

### Graph 5: Heatmap of firearm counts across different states in the US

In [8]:
### Heatmap for the number of firearm counts across States and Years.
alt.Chart(firearm).mark_bar().transform_fold(
    ['Alabama', 'Alaska',
       'Arizona', 'Arkansas', 'California', 'Colorado', 'Connecticut',
       'Delaware', 'DistrictofColumbia', 'Florida', 'Georgia', 'Hawaii',
       'Idaho', 'Illinois', 'Indiana', 'Iowa', 'Kansas', 'Kentucky',
       'Louisiana', 'Maine', 'Maryland', 'Massachusetts', 'Michigan',
       'Minnesota', 'Mississippi', 'Missouri', 'Montana', 'Nebraska', 'Nevada',
       'NewHampshire', 'NewJersey', 'NewMexico', 'NewYork', 'NorthCarolina',
       'NorthDakota', 'Ohio', 'Oklahoma', 'Oregon', 'Pennsylvania',
       'RhodeIsland', 'SouthCarolina', 'SouthDakota', 'Tennessee', 'Texas',
       'Utah', 'Vermont', 'Virginia', 'Washington', 'WestVirginia',
       'Wisconsin', 'Wyoming'],as_=['state','count']
).encode(
    alt.X('Years:O', 
          
          axis=alt.Axis(labelAngle=-45,
                        )),
    alt.Y('state:N'),
    alt.Color('count:Q'),
    tooltip = ['Years:O','state:N','count:Q']
).properties(
    title='Firearm Deaths Across States: 1981-2021'
)

alt.Chart(...)

#### Observations:
- Analyzing the color contrast, we can observe a significant decline in firearm-related deaths in California since 1993.
- The graph makes it clear that the top three states with the highest incidence of firearm-related deaths are Texas, California, and Florida.

However, obtaining counts for the remaining states is challenging. Streamlining the number of states for this exploration or plotting them on a map would help in a more focused analysis of the data.

Note: While the current graph may not provide the most accurate depiction of this relationship, it serves as an initial exploration. Further improvements can be made to refine and enhance the visualization.

#### Effectiveness of the Graph:

The heatmap titled "Firearm Deaths Across States: 1981-2021" provides a visual representation of firearm deaths across different states over the years from 1981 to 2021. Each cell in the heatmap represents the number of firearm deaths in a specific state for a particular year. The x-axis shows the years, the y-axis shows the states, and the color intensity indicates the count of firearm deaths. This heatmap allows for easy comparison of firearm deaths between states and over the specified time range. The color variation helps to identify states with higher or lower firearm death counts in a visually intuitive manner.

# Preliminary Graphs

### 1. Identifying States with maximum and minumum firearm death count and compare it with Washington State
The aim is to determine the state with the highest and lowest firearm-related deaths and understand how these figures relate to the broader spectrum of causes of death during the respective periods. Additionally, will compare this data to the state of Washington to provide a comparative perspective.

#### Let's find the states with the minimum and maximum firearm deaths from 1981 to 2021

In [9]:
state_sum = firearm[['Alabama', 'Alaska',
       'Arizona', 'Arkansas', 'California', 'Colorado', 'Connecticut',
       'Delaware', 'DistrictofColumbia', 'Florida', 'Georgia', 'Hawaii',
       'Idaho', 'Illinois', 'Indiana', 'Iowa', 'Kansas', 'Kentucky',
       'Louisiana', 'Maine', 'Maryland', 'Massachusetts', 'Michigan',
       'Minnesota', 'Mississippi', 'Missouri', 'Montana', 'Nebraska', 'Nevada',
       'NewHampshire', 'NewJersey', 'NewMexico', 'NewYork', 'NorthCarolina',
       'NorthDakota', 'Ohio', 'Oklahoma', 'Oregon', 'Pennsylvania',
       'RhodeIsland', 'SouthCarolina', 'SouthDakota', 'Tennessee', 'Texas',
       'Utah', 'Vermont', 'Virginia', 'Washington', 'WestVirginia',
       'Wisconsin', 'Wyoming']].sum()

In [10]:

print('The State with minimum number of Firearm Deaths in USA: ',state_sum[state_sum == state_sum.min()])
print('The State with maximum number of Firearm Deaths in USA: ',state_sum[state_sum == state_sum.max()])

The State with minimum number of Firearm Deaths in USA:  RhodeIsland    2005
dtype: int64
The State with maximum number of Firearm Deaths in USA:  California    150943
dtype: int64


#### Rhode Island holds the record for the lowest number of firearm-related deaths, while California leads with the highest count.

In [11]:
## Cause of Firearm Deaths Over Time
cause_firearm_time = alt.Chart(firearm).transform_fold(
    ['California','RhodeIsland','Washington'],
    as_=['State','Count']
).mark_line( ).encode(
    alt.X('Years:O',
           axis = alt.Axis(labelAngle=0,title='Year',labelFontSize = 12,
                         titleFontSize = 15,
                         values=[1980, 1985, 1990, 1995, 2000, 2005, 2010, 2015, 2020])),
    alt.Y('Count:Q',
           axis = alt.Axis(title='Number Of Firearm Deaths',labelFontSize = 15,
                         titleFontSize = 15)),
    alt.Color('State:N',legend=None),
    text=alt.Text('State:N'),
    tooltip =['State:N','Years:O','Count:Q']
    
).properties(
    title = alt.Title(text='Firearm Death Count By State',fontSize=20)
).properties(
    width=700,
    height=300,
    title = alt.Title(text='Firearm Death Total, California, Washington and RhodeIsland',fontSize=20)
)


labels = alt.Chart(firearm).transform_fold(
   ['California','RhodeIsland','Washington'],
    as_=['State','Count']
).mark_text(align='left', dx=6,dy = -15,size = 13).encode(
    alt.X('Years:O', aggregate='max'),
    alt.Y('Count:Q', aggregate={'argmin': 'Count'}),
    alt.Text('State:N'),
    alt.Color('State:N', legend=None)
)


line_chart_state = alt.layer(cause_firearm_time , labels).resolve_scale(color='independent')


#### Change the plot
total_state_min_max = alt.Chart(firearm).mark_bar().transform_fold(
   ['California','RhodeIsland','Washington'],
    as_=['State','Count']
).transform_aggregate(
    sum_count='sum(Count)',
    groupby=['State']
).encode(
    alt.X('sum_count:Q',
          axis =alt.Axis(title='Number of Firearm Deaths',
                         labelFontSize = 15,
                         titleFontSize = 15)),
    alt.Y('State:N', 
          axis = alt.Axis(labelFontSize=15,
                          title='State',titleFontSize=15),
          sort = ['California','Washington','RhodeIsland']),  # Sorting age groups in descending order
 
    tooltip=['sum_count:Q', 'State:N'],
    
).properties(
    width=700,
    height=300,
    title = alt.Title(text='Firearm Death Total, California, Washington and RhodeIsland',fontSize=20)
)




wahsington_cal_rhode = alt.vconcat(total_state_min_max,line_chart_state)
wahsington_cal_rhode.save('wahsington_cal_rhode.html')
cause_firearm_time

alt.Chart(...)

In [22]:
cause_firearm_time

alt.Chart(...)

#### Observation:

We can see that the RhodeIsland has very few firearm deaths across years. It may be due to the fact that the Rhode Island is really very small state. 
- We can also see that the the california had maximum numebr of deaths during 1993. Whereas the rate of firearm deaths for Washington and Rhode Island seems constant. there a rise in the firearm deaths in Washington after 2016 but it's not that high. 

- We can also see that the Total Firearm deaths across all years is around 1,500,000 but the total count of Rhode Island is significantly small:2005 and even the count of califronia is 150943. 


The line graph helps in visuallizing the trends over time. We can see that the legends for showing the color is also very nice. One doen't have to make lot of eye moment to understand the data. 

#### Effectiveness of the Graph:

The graphic titled "Firearm Death Count By State" presents the firearm death count over time for the states of California, RhodeIsland, and Washington. Each line represents the trend of firearm deaths for a specific state from 1981 to 2021. The x-axis shows the years, the y-axis shows the count of firearm deaths, and different colors represent each state. The graph effectively provides a comparison of firearm death counts across these states, helping to identify trends and variations over the specified time range. The addition of labels at end of the line enhances the understanding of the data. 

In addition to that,the bar chart shows the comparison of the total firearm deaths in these states, making it easier to observe and analyze variations in counts between the states. Overall, the graphic is doing justice in illustrating the variations in firearm deaths for the selected states over the years.


Let's shift our focus to the Washington state and help ourselves to understand the situation and try to answer the following question

### 2.  Let's see the minimum and maximum number of firearm deaths in Washington and when does that happen?? How was the firearm rate during that time across male and female? What was the trend during that across multiple age groups? 

### Let's find the dates when the crime was really low and high in Washington State

In [12]:
print('The firearm deaths were at their minimum in Washington during the year:  ',firearm[firearm.Washington==firearm.Washington.min()][['Years']])
print('The firearm deaths were at their maximum in Washington during the year: ',firearm[firearm.Washington==firearm.Washington.max()][['Years']])


The firearm deaths were at their minimum in Washington during the year:      Years
3   1984
The firearm deaths were at their maximum in Washington during the year:      Years
40   2021


#### Maximum and the minimum number of deaths in Washington occured during 2021 and 1984. Let's explore the stated relationships and see how does that help in answering our question?

In [13]:
washington_firearm_time = alt.Chart(firearm).mark_line().encode(
    alt.X('Years:O', axis=alt.Axis(labelAngle=-45, title='Year')),
    alt.Y('Washington:Q', axis=alt.Axis(title='Number Of Firearm Deaths')),
    tooltip =['Years:O','Washington:Q']
).properties(
    title=alt.Title(text='Firearm Death Count', fontSize=16)
)


min_point = washington_firearm_time.mark_point(color='blue').transform_filter(
    alt.datum.Washington == min(firearm['Washington'])
).encode(
    alt.Text('Washington:Q', format='.0f'),
    alt.ColorValue('#1f77b4')
).mark_text(align='left', dx=5,size = 15)

max_point = washington_firearm_time.mark_point(color='orange').transform_filter(
    alt.datum.Washington == max(firearm['Washington'])
).encode(
    alt.Text('Washington:Q', format='.0f'),
    alt.ColorValue('#ff7f0e')
).mark_text(align='left', dx=5, dy=10,size = 17)


washington_firearm_time = (washington_firearm_time + min_point + max_point)


years_age_washington_min_max = alt.Chart(firearm).transform_filter(
     (alt.datum.Years == 1984) | (alt.datum.Years==2021)
).transform_fold(
    ['Under19', '20-34', '35-49', '50-64', '65+'],
    as_=['Age', 'Count']
).mark_bar().encode(
    x=alt.X('Age:O',
            axis=alt.Axis(labelAngle=-45),sort = ['Under19', '20-34', '35-49', '50-64', '65+']),
    xOffset='Years:N',
    y=alt.Y('Count:Q'),
    color=alt.Color('Years:N'),
).properties(
    width =350,
    height=300,
    title=alt.Title(text='Comparison By Age: 1984 and 2021', fontSize=16)
)

years_gender_washington_min_max  = alt.Chart(firearm).transform_filter(
     (alt.datum.Years == 1984) | (alt.datum.Years==2021)
).transform_fold(
    ['Male','Female'],
    as_ =['Gender','Count']
).mark_bar().encode(
    y=alt.Y('Gender:N'),
    yOffset='Years:N',
    x=alt.X('Count:Q'),
    color=alt.Color('Years:N',
                    scale=alt.Scale( scheme='category10')),
).properties(
    width =350,
    height=300,
    title=alt.Title(text='Comparison By Gender: 1984 and 2021', fontSize=16)
)



concatenated_chart = alt.vconcat(washington_firearm_time,
                                 alt.hconcat(years_age_washington_min_max,
                                             years_gender_washington_min_max))


(concatenated_chart.properties(
    title=alt.Title(text='Firearm Death Analysis for Washington', fontSize=20)
).configure_view(
    stroke='transparent'
).configure_legend(
    labelFontSize=15,  
    titleFontSize=15  
))

alt.VConcatChart(...)

#### Observations

- The minimum and maximum number of firearm deaths in Washington were 430 and 896, respectively.
- Across each age group, there was a noticeable increase in the number of firearm deaths.
- The 20-34 age group showed the highest number of firearm deaths.
- The disparity between male and female firearm deaths is significant. The difference in firearm deaths for male is otably higher compared to females.


#### Effectiveness of the Graph
The graph titled "Firearm Death Analysis for Washington" comprises multiple subplots, offering a comprehensive view of firearm deaths in Washington. The main line chart presents firearm death counts for Washington over the years, highlighting the minimum and maximum points. Additional subplots break down the data by age groups and gender for specific years, providing valuable insights into the distribution of firearm deaths. The use of color (orange for 2021 and blue for 1984) helps in understanding the difference between the count for both the years and tooltips also enhances the interpretability of the data. The layout is clean and informative, effectively conveying the relevant information for understanding firearm deaths in Washington. The legend and title sizes are appropriately adjusted for clarity and aesthetics. Overall, the graphic effectively addresses the question of firearm deaths in Washington, providing valuable insights and visualizations.


### Graph3: How Have Firearm Deaths and Causes Evolved from 1981 to 2021?

In [14]:
bar_min_max = alt.Chart(firearm).transform_filter(
    (alt.datum.Years == max(firearm['Years'])) |
    (alt.datum.Years == min(firearm['Years']))
).mark_bar().encode(
    alt.X('Years:O',axis=alt.Axis(labelAngle=0)),
    alt.Y('Firearmdeaths:Q')
).properties(
    width=300,
    height=400,
    title='Firearm Deaths Count: 1981 and 2021'
)


base_chart = alt.Chart(firearm).transform_filter(
   (alt.datum.Years == max(firearm['Years'])) |
    (alt.datum.Years == min(firearm['Years']))
).transform_fold(
     ['Homicide', 'Suicide', 'Unintentional',
       'LegalIntervention', 'Undetermined'],
    as_=['Cause', 'Count']
).mark_line(opacity=0.8, interpolate='basis').encode(
    alt.X('Years:O', title='Year',axis=alt.Axis(labelAngle=0)),
    alt.Y('Count:Q', title='Count of Firearm Deaths'),
    alt.Color('Cause:N', title='Cause of Death', scale=alt.Scale(scheme='category10'))
).properties(
    width=300,
    height=400,
    title='Firearm Deaths by Cause: 1981 and 2021 Year'
)

change_1981_2021 = alt.hconcat(bar_min_max,base_chart)



change_1981_2021.save('change_1981_2021.html')
change_1981_2021

alt.HConcatChart(...)

#### Observations
- The incidence of firearm deaths has significantly risen between 1981 and 2021.
- Noticeably, both Homicide and Suicide cases have shown an increase during this period.
- Conversely, there has been a decline in Unintentional cases over time.


#### Effectiveness of the Graph
This layout effectively compares firearm death counts for the years 1981 and 2021. The left bar chart shows the total firearm deaths for these years, while the right chart provides a detailed breakdown of firearm deaths by cause and year. The line chart presents data for various causes of death, allowing for a clear comparison of trends. The colors and layout contribute to the visual appeal and understanding of the data. The title appropriately conveys the purpose of the graph.

## Refined Graphics

### Iteration 1

- Add space between the combined graphs

In [15]:
bar_min_max = alt.Chart(firearm).transform_filter(
    (alt.datum.Years == max(firearm['Years'])) |
    (alt.datum.Years == min(firearm['Years']))
).mark_bar().encode(
    alt.X('Years:O'),
    alt.Y('Firearmdeaths:Q')
).properties(
    width=300,
    height=400,
    title= 'Firearm Deaths Count: 1981 and 2021'
)

base_chart = alt.Chart(firearm).transform_filter(
   (alt.datum.Years == max(firearm['Years'])) |
    (alt.datum.Years == min(firearm['Years']))
).transform_fold(
     ['Homicide', 'Suicide', 'Unintentional',
       'LegalIntervention', 'Undetermined'],
    as_=['Cause', 'Count']
).mark_line(opacity=0.8, interpolate='basis').encode(
    alt.X('Years:O', title='Year'),
    alt.Y('Count:Q', title='Count of Firearm Deaths'),
    alt.Color('Cause:N', title='Cause of Death', scale=alt.Scale(scheme='category10'))
).properties(
    width=300,
    height=400,
    title='Change In Count of Firearm Deaths : 1981 and 2021'
)





combined_chart = alt.hconcat(bar_min_max, base_chart, spacing=70).properties()

combined_chart


alt.HConcatChart(...)

### Iteration 2

It's hard to read the year in the previous graph. 

- Rotate the X-axis label. It's hard to read
- Increase the size of labels

In [16]:
bar_min_max = alt.Chart(firearm).transform_filter(
    (alt.datum.Years == max(firearm['Years'])) |
    (alt.datum.Years == min(firearm['Years']))
).mark_bar().encode(
    alt.X('Years:O',
          title='Year',
          axis=alt.Axis(labelAngle=0,
                        labelFontSize =15,
                        titleFontSize=14)),
    alt.Y('Firearmdeaths:Q',
          title='Count of Firearm Deaths', 
          axis = alt.Axis(labelFontSize=13,titleFontSize=14))
).properties(
    width=300,
    height=400,
    title= 'Firearm Deaths Count: 1981 and 2021'
)

base_chart = alt.Chart(firearm).transform_filter(
   (alt.datum.Years == max(firearm['Years'])) |
    (alt.datum.Years == min(firearm['Years']))
).transform_fold(
     ['Homicide', 'Suicide', 'Unintentional',
       'LegalIntervention', 'Undetermined'],
    as_=['Cause', 'Count']
).mark_line(opacity=0.8, interpolate='basis').encode(
    alt.X('Years:O', title='Year',axis=alt.Axis(labelAngle=0,
                        labelFontSize =15,
                        titleFontSize=14)),
    alt.Y('Count:Q', 
          title='Count of Firearm Deaths', 
          axis = alt.Axis(labelFontSize=13,titleFontSize=14)),
    alt.Color('Cause:N', title='Cause of Death', scale=alt.Scale(scheme='category10'))
).properties(
    width=300,
    height=400,
    title='Change In Count of Firearm Deaths : 1981 and 2021'
)





combined_chart = alt.hconcat(bar_min_max, base_chart, spacing=70).properties(
 

)

# Show the combined chart
combined_chart


alt.HConcatChart(...)

### Iteration 3

We can enhance the graph by incorporating tooltips, making it easier for viewers to interpret the data points when hovering over the graph.

In [17]:
bar_min_max = alt.Chart(firearm).transform_filter(
    (alt.datum.Years == max(firearm['Years'])) |
    (alt.datum.Years == min(firearm['Years']))
).mark_bar().encode(
    alt.X('Years:O',
          title='Year',
          axis=alt.Axis(labelAngle=0,
                        labelFontSize =15,
                        titleFontSize=14)),
    alt.Y('Firearmdeaths:Q',
          title='Count of Firearm Deaths', 
          axis = alt.Axis(labelFontSize=13,titleFontSize=14)),
    tooltip = ['Years:O','Firearmdeaths:Q']
).properties(
    width=300,
    height=400,
    title= 'Firearm Deaths Count: 1981 and 2021'
)

base_chart = alt.Chart(firearm).transform_filter(
   (alt.datum.Years == max(firearm['Years'])) |
    (alt.datum.Years == min(firearm['Years']))
).transform_fold(
     ['Homicide', 'Suicide', 'Unintentional',
       'LegalIntervention', 'Undetermined'],
    as_=['Cause', 'Count']
).mark_line(opacity=0.8, interpolate='basis').encode(
    alt.X('Years:O',
          title='Year',
          axis=alt.Axis(labelAngle=0,
                        labelFontSize =15,
                        titleFontSize=14)),
    alt.Y('Count:Q', 
          title='Count of Firearm Deaths', 
          axis = alt.Axis(labelFontSize=13,titleFontSize=14)),
    alt.Color('Cause:N', title='Cause of Death', scale=alt.Scale(scheme='category10')),
    tooltip = ['Years:O','Cause:N','Count:Q']
).properties(
    width=300,
    height=400,
    title='Change In Count of Firearm Deaths : 1981 and 2021'
)





combined_chart = alt.hconcat(bar_min_max, base_chart, spacing=70).properties()


combined_chart


alt.HConcatChart(...)

### Iteration 4

Increase the size of the legend as it's hard to read the labels

In [18]:
bar_min_max = alt.Chart(firearm).transform_filter(
    (alt.datum.Years == max(firearm['Years'])) |
    (alt.datum.Years == min(firearm['Years']))
).mark_bar().encode(
    alt.X('Years:O',
          title='Year',
          axis=alt.Axis(labelAngle=0,
                        labelFontSize =15,
                        titleFontSize=14)),
    alt.Y('Firearmdeaths:Q',
          title='Count of Firearm Deaths', 
          axis = alt.Axis(labelFontSize=13,titleFontSize=14)),
    tooltip = ['Years:O','Firearmdeaths:Q']
).properties(
    width=300,
    height=400,
    title= 'Firearm Deaths Count: 1981 and 2021'
)

base_chart = alt.Chart(firearm).transform_filter(
   (alt.datum.Years == max(firearm['Years'])) |
    (alt.datum.Years == min(firearm['Years']))
).transform_fold(
     ['Homicide', 'Suicide', 'Unintentional',
       'LegalIntervention', 'Undetermined'],
    as_=['Cause', 'Count']
).mark_line(opacity=0.8, interpolate='basis').encode(
    alt.X('Years:O',
          title='Year',
          axis=alt.Axis(labelAngle=0,
                        labelFontSize =15,
                        titleFontSize=14)),
    alt.Y('Count:Q', 
          title='Count of Firearm Deaths', 
          axis = alt.Axis(labelFontSize=13,titleFontSize=14)),
    alt.Color('Cause:N', title='Cause of Death', scale=alt.Scale(scheme='category10')),
    tooltip = ['Years:O','Cause:N','Count:Q']
).properties(
    width=300,
    height=400,
    title='Change In Count of Firearm Deaths : 1981 and 2021'
)

combined_chart = alt.hconcat(bar_min_max, base_chart, spacing=70).properties(
    config={'legend': {'labelFontSize': 14, 'titleFontSize': 15}}
)

# Show the combined chart
combined_chart


alt.HConcatChart(...)

### Iteration 5

Add information about the data source : Increase credibility and transparency of the dataset.

In [19]:
bar_min_max = alt.Chart(firearm).transform_filter(
    (alt.datum.Years == max(firearm['Years'])) |
    (alt.datum.Years == min(firearm['Years']))
).mark_bar().encode(
    alt.X('Years:O',
          title='Year',
          axis=alt.Axis(labelAngle=0,
                        labelFontSize =15,
                        titleFontSize=14)),
    alt.Y('Firearmdeaths:Q',
          title='Count of Firearm Deaths', 
          axis = alt.Axis(labelFontSize=13,titleFontSize=14)),
    tooltip = ['Years:O','Firearmdeaths:Q']
).properties(
    width=300,
    height=400,
    title= 'Firearm Deaths Count: 1981 and 2021'
)

base_chart = alt.Chart(firearm).transform_filter(
   (alt.datum.Years == max(firearm['Years'])) |
    (alt.datum.Years == min(firearm['Years']))
).transform_fold(
     ['Homicide', 'Suicide', 'Unintentional',
       'LegalIntervention', 'Undetermined'],
    as_=['Cause', 'Count']
).mark_line(opacity=0.8, interpolate='basis').encode(
    alt.X('Years:O',
          title='Year',
          axis=alt.Axis(labelAngle=0,
                        labelFontSize =15,
                        titleFontSize=14)),
    alt.Y('Count:Q', 
          title='Count of Firearm Deaths', 
          axis = alt.Axis(labelFontSize=13,titleFontSize=14)),
    alt.Color('Cause:N', title='Cause of Death', scale=alt.Scale(scheme='category10')),
    tooltip = ['Years:O','Cause:N','Count:Q']
).properties(
    width=300,
    height=400,
    title='Change In Count of Firearm Deaths : 1981 and 2021'
)

combined_chart = alt.hconcat(bar_min_max, base_chart, spacing=50)



source_text = alt.Chart(pd.DataFrame({'source': ['Source:https://usafacts.org/data/topics/security-safety/crime-and-justice/firearms/firearm-deaths/']})).mark_text(
    fontSize=12,
    text='source',
    dx=350
).encode(
    text='source:N'
)


alt.vconcat(combined_chart, source_text).properties(
    config={'legend': {'labelFontSize': 14, 
                       'titleFontSize': 15 ,
                       'padding': 0,
                      'labelOffset':10,
                       'offset':0.1
                      }}
)

alt.VConcatChart(...)

### Iteration 6

Add common title to both the graphs

In [20]:
bar_min_max = alt.Chart(firearm).transform_filter(
    (alt.datum.Years == max(firearm['Years'])) |
    (alt.datum.Years == min(firearm['Years']))
).mark_bar().encode(
   alt.X('Years:O',
          title='Year',
          axis=alt.Axis(labelAngle=0,
                        labelFontSize =15,
                        titleFontSize=14)),
    alt.Y('Firearmdeaths:Q',
          title='Count of Firearm Deaths', 
          axis = alt.Axis(labelFontSize=13,titleFontSize=14)),
    tooltip = ['Years:O','Firearmdeaths:Q']
).properties(
    width=300,
    height=400,
    title = 'Firearm Deaths Count'
)

base_chart = alt.Chart(firearm).transform_filter(
   (alt.datum.Years == max(firearm['Years'])) |
    (alt.datum.Years == min(firearm['Years']))
).transform_fold(
     ['Homicide', 'Suicide', 'Unintentional',
       'LegalIntervention', 'Undetermined'],
    as_=['Cause', 'Count']
).mark_line(opacity=0.8, interpolate='basis').encode(
   alt.X('Years:O',
          title='Year',
          axis=alt.Axis(labelAngle=0,
                        labelFontSize =15,
                        titleFontSize=14)),
    alt.Y('Count:Q', 
          title='Count of Firearm Deaths', 
          axis = alt.Axis(labelFontSize=13,titleFontSize=14)),
    alt.Color('Cause:N', title='Cause of Death', scale=alt.Scale(scheme='category10')),
    tooltip = ['Years:O','Cause:N','Count:Q']
).properties(
    width=300,
    height=400,
    title='Change In Count of Firearm Deaths',
    
)

combined_chart = alt.hconcat(bar_min_max, base_chart, spacing=50)

source_text = alt.Chart(pd.DataFrame({'source': ['Source:https://usafacts.org/data/topics/security-safety/crime-and-justice/firearms/firearm-deaths/']})).mark_text(
    fontSize=12,
    text='source',
    dx=350
).encode(
    text='source:N'
)


alt.vconcat(combined_chart, source_text).properties(
    config={'legend': {'labelFontSize': 14, 
                       'titleFontSize': 15 ,
                       'padding': 0,
                       'labelOffset':10,
                       'offset':0.1
                      }},
    title = {
        "text" : ['Comparative Analysis Of Firearm Deaths: 1981 and 2021 Year'],
        #"subtitle":['1981 and 2021 Year'],
        "dx":180,
        "fontSize":20,
        "offset":20,
        "subtitlePadding":10,
        "subtitleFontSize":18
    
    }
)




alt.VConcatChart(...)

### Iteration 7
- Title of the individual graph: Increase the size of it

In [21]:
bar_min_max = alt.Chart(firearm).transform_filter(
    (alt.datum.Years == max(firearm['Years'])) |
    (alt.datum.Years == min(firearm['Years']))
).mark_bar().encode(
   alt.X('Years:O',
          title='Year',
          axis=alt.Axis(labelAngle=0,
                        labelFontSize =15,
                        titleFontSize=14)),
    alt.Y('Firearmdeaths:Q',
          title='Count of Firearm Deaths', 
          axis = alt.Axis(labelFontSize=13,titleFontSize=14)),
    tooltip = ['Years:O','Firearmdeaths:Q']
).properties(
    width=300,
    height=400,
    #title= 'Firearm Deaths Count'
    title = {
        "text" : ['Firearm Deaths Count'],
        "fontSize":16}
)

base_chart = alt.Chart(firearm).transform_filter(
   (alt.datum.Years == max(firearm['Years'])) |
    (alt.datum.Years == min(firearm['Years']))
).transform_fold(
     ['Homicide', 'Suicide', 'Unintentional',
       'LegalIntervention', 'Undetermined'],
    as_=['Cause', 'Count']
).mark_line(opacity=0.8, interpolate='basis').encode(
   alt.X('Years:O',
          title='Year',
          axis=alt.Axis(labelAngle=0,
                        labelFontSize =15,
                        titleFontSize=14)),
    alt.Y('Count:Q', 
          title='Count of Firearm Deaths', 
          axis = alt.Axis(labelFontSize=13,titleFontSize=14)),
    alt.Color('Cause:N', title='Cause of Death', scale=alt.Scale(scheme='category10')),
    tooltip = ['Years:O','Cause:N','Count:Q']
).properties(
    width=300,
    height=400,    
    title = {
        "text" : ['Change In Count of Firearm Deaths'],
        "fontSize":16}
    
)

combined_chart = alt.hconcat(bar_min_max, base_chart, spacing=50)

source_text = alt.Chart(pd.DataFrame({'source': ['Source:https://usafacts.org/data/topics/security-safety/crime-and-justice/firearms/firearm-deaths/']})).mark_text(
    fontSize=12,
    text='source',
    dx=350
).encode(
    text='source:N'
)


alt.vconcat(combined_chart, source_text).properties(
    config={'legend': {'labelFontSize': 14, 
                       'titleFontSize': 15 ,
                       'padding': 0,
                       'labelOffset':10,
                       'offset':0.1
                      }},
   title = {
        "text" : ['Comparative Analysis Of Firearm Deaths: 1981 and 2021 Year'],
        #"subtitle":['1981 and 2021 Year'],
        "dx":180,
        "fontSize":20,
        "offset":20,
        "subtitlePadding":10,
        "subtitleFontSize":18
    
    }
)




alt.VConcatChart(...)